File: algorithms/Minimum_Difference_Between_Largest_and_Smallest_Value_in_Three_Moves-medium/solution1.cpp
class Solution {
public:
    int minDifference(vector<int>& nums) {
        if (nums.size() <= 4) return 0;
        int maxq[4];
        int minq[4];
        fill_n(maxq, 4, INT_MIN);
        fill_n(minq, 4, INT_MAX);
        for (int n : nums) {
            getMax(maxq, n);
            getMin(minq, n);
        }
        int res = INT_MAX;
        for (int i = 0; i < 4; i++) {
            res = min(res, maxq[i] - minq[3-i]);
        }
        return res;
    }
    
    void getMax(int maxq[4], int n) {
        for (int i = 0; i < 4; i++) {
            if (maxq[i] < n) {
                int tmp = n;
                n = maxq[i];
                maxq[i] = tmp;
            }
        }
    }
    
    void getMin(int minq[4], int n) {
        for (int i = 0; i < 4; i++) {
            if (minq[i] > n) {
                int tmp = n;
                n = minq[i];
                minq[i] = tmp;
            }
        }
    }
};

File: algorithms/Count_Square_Submatrices_with_All_Ones-medium/solution1.java
class Solution {
    public int countSquares(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (matrix[i][j] == 1) {
                    int min = Math.min(matrix[i-1][j-1], matrix[i-1][j]);
                    min = Math.min(matrix[i][j-1], min);
                    matrix[i][j] = min + 1;
                }
            }
        }
        int res = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {        
                res += matrix[i][j];
            }
        }
        return res;
    }
}
